### ISSUE TITLE
Finding ideal number of layers via recursion and entropy

### ISSUE THREAD

**Issue by yashwantram97** [The-School-of-AI/LLM#337]
# Determine Optimal Model Depth via Recursive Training & Entropy Analysis

## Summary
Find the maximum trainable depth for our model architecture by progressively growing depth and monitoring training stability through entropy metrics and loss curves.

## Problem
When scaling model depth (layers), we need to know:
- How deep can we go before training becomes unstable?
- What's the optimal depth for our compute budget?
- When does adding more layers stop helping?

### 1. Progressive Depth Growth
Start shallow, incrementally add layers, monitor for:
- Loss convergence
- Gradient flow stability  
- Training dynamics breakdown

### 2. Entropy-Based Stability Signals

**Layer-wise entropy tracking:**
```python
# Monitor entropy at each layer during training
H(layer_i) = -Σ p(x) log p(x)

# Warning signs:
# - Entropy collapse (→0): representations becoming deterministic
# - Entropy explosion (→∞): loss of information structure
# - Entropy plateaus: diminishing returns from depth
```

### 3. Recursive Validation Loop

For each depth candidate:
1. Train for N steps
2. Measure: loss curve, gradient norms, activation entropy
3. Check: Is training stable? Is loss improving?
4. If stable → try deeper
5. If unstable → found depth limit

## Warning Signals

**Stop growing depth when:**
- Training becomes unstable (loss spikes, NaN gradients)
- Entropy collapses or explodes
- Marginal loss improvement < 1%
- Gradient vanishing/explosion
- Activation rank collapse

Ref paper:: https://arxiv.org/pdf/2510.25741v2
